8342075: HttpClient: improve HTTP/2 flow control checks

alexeybakhtin opened the pull request:
Please review the backport of JDK-8342075

Almost clean backport except for the import section of src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java

Requires JDK-8317808 [1] as a prerequisite to fixing the hanging of the new tests.

All jtreg tests for java.net.httpclient are passed

[1] - https://github.com/openjdk/jdk21u-dev/pull/1426
[2] - https://github.com/openjdk/jdk/commit/720b44648bcff997278af92746f942b2425298a5#diff-c18c2a070038409f25c08e13c174a95f33840e2deddcc29f6d8306de572a618b

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] [JDK-8342075](https://bugs.openjdk.org/browse/JDK-8342075) needs maintainer approval
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8350602](https://bugs.openjdk.org/browse/JDK-8350602) to be approved

### Issues
 * [JDK-8342075](https://bugs.openjdk.org/browse/JDK-8342075): HttpClient: improve HTTP/2 flow control checks (**Bug** - P4 - Approved)
 * [JDK-8350602](https://bugs.openjdk.org/browse/JDK-8350602): HttpClient: improve HTTP/2 flow control checks (**CSR**)


### Reviewers
 * [Andrew John Hughes](https://openjdk.org/census#andrew) (@gnu-andrew - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk21u-dev.git pull/1428/head:pull/1428` \
`$ git checkout pull/1428`

Update a local copy of the PR: \
`$ git checkout pull/1428` \
`$ git pull https://git.openjdk.org/jdk21u-dev.git pull/1428/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1428`

View PR using the GUI difftool: \
`$ git pr show -t 1428`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk21u-dev/pull/1428.diff">https://git.openjdk.org/jdk21u-dev/pull/1428.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk21u-dev/pull/1428#issuecomment-2675636583)
</details>

alexeybakhtin commented:
CSR is submitted : https://bugs.openjdk.org/browse/JDK-8350611

gnu-andrew reviewed (commented):
Code backport looks good (and matches what I started doing myself with resolving the import).

With the test changes, was there a reason for abandoning #1416 & #1417 and including the test changes in this patch? It looks like they are both in 21.0.8-oracle

alexeybakhtin commented:
At the time of backport creation, JDK-8315181 was not integrated into Oracle 21.0.8 yet. This is why I decided to make a patch without it, especially because JDK-8315181 is required for tests only.
Now JDK-8315181 is integrated into Oracle and I reopened https://github.com/openjdk/jdk21u-dev/pull/1416 & https://github.com/openjdk/jdk21u-dev/pull/1417 backports.
I will rebase this patch as soon as https://github.com/openjdk/jdk21u-dev/pull/1417 is accepted

alexeybakhtin commented:
@gnu-andrew 
I have rebased the PR to the latest master. 
The backport is almost clean now, except for the import section in Http2Connection.java. This section was added by JDK-8310330 in JDK22 but not backported.
I have removed test changes from the initial description as soon as they are in the master already.
All related jtreg tests are passed

gnu-andrew commented:
> CSR [JDK-8350602](https://bugs.openjdk.org/browse/JDK-8350602) for JDK21 is approaved

Yes, I just linked that one to [JDK-8350610](https://bugs.openjdk.org/browse/JDK-8350610) in place of the withdrawn one.

gnu-andrew reviewed (approved):
Thanks for handling the pre-requisites.  Patch looks good. Please apply for approval.

gnu-andrew commented:
/approve yes

alexeybakhtin commented:
/integrate
